bundles/org.dataflowanalysis.analysis.dfd/src/org/dataflowanalysis/analysis/dfd/core/DFDCyclicTransposeFlowGraphFinder.java

```diff
@@ -0,0 +1,222 @@
+import org.dataflowanalysis.dfd.datadictionary.DataDictionary;
+import org.dataflowanalysis.dfd.datadictionary.Pin;
+import org.dataflowanalysis.dfd.dataflowdiagram.DataFlowDiagram;
+import org.dataflowanalysis.dfd.dataflowdiagram.Flow;
+import org.dataflowanalysis.dfd.dataflowdiagram.Node;
+import org.eclipse.emf.ecore.util.EcoreUtil;
+
+/**
+ * The DFDTransposeFlowGraphFinder determines all transpose flow graphs contained in a model, allowing cycles
+ */
+public class DFDCyclicTransposeFlowGraphFinder extends DFDTransposeFlowGraphFinder {
+    private final Logger logger = Logger.getLogger(DFDCyclicTransposeFlowGraphFinder.class);
+    private boolean hasCycles = false;
+
+    /***
+     * The DFDTransposeFlowGraphFinder determines all transpose flow graphs contained in a model, allowing cycles
+     */
+    public DFDCyclicTransposeFlowGraphFinder(DFDResourceProvider resourceProvider) {
+        super(resourceProvider);
+    }
+
+    /***
+     * The DFDTransposeFlowGraphFinder determines all transpose flow graphs contained in a model, allowing cycles
+     * @param dataDictionary
+     * @param dataFlowDiagram
+     */
+    public DFDCyclicTransposeFlowGraphFinder(DataDictionary dataDictionary, DataFlowDiagram dataFlowDiagram) {
+        super(dataDictionary, dataFlowDiagram);
+    }
+
+    /**
+     * Finds all transpose flow graphs in a dataflowdiagram model instance
+     * @return Returns a list of all transpose flow graphs
+     */
+
+    @Override
+    public List<? extends AbstractTransposeFlowGraph> findTransposeFlowGraphs(List<?> sinkNodes, List<?> sourceNodes) {
+        List<Node> potentialSinks = sinkNodes.stream()
+                .filter(Node.class::isInstance)
+                .map(Node.class::cast)
+                .toList();
+        List<Node> sources = sourceNodes.stream()
+                .filter(Node.class::isInstance)
+                .map(Node.class::cast)
+                .toList();
+        List<DFDTransposeFlowGraph> transposeFlowGraphs = new ArrayList<>();
+
+        for (Node endNode : potentialSinks) {
+            List<DFDVertex> sinks = determineSinks(new DFDVertex(endNode, new HashMap<>(), new HashMap<>()), endNode.getBehaviour()
+                    .getInPin(), sources);
+            if (!sourceNodes.isEmpty()) {
+                sinks = sinks.stream()
+                        .filter(it -> new DFDTransposeFlowGraph(it).getVertices()
+                                .stream()
+                                .filter(DFDVertex.class::isInstance)
+                                .map(DFDVertex.class::cast)
+                                .anyMatch(vertex -> sources.contains(vertex.getReferencedElement())))
+                        .toList();
+            }
+            sinks.forEach(sink -> sink.unify(new HashSet<>()));
+            sinks.forEach(sink -> transposeFlowGraphs.add(new DFDTransposeFlowGraph(sink)));
+        }
+        return transposeFlowGraphs;
+    }
+
+    public Boolean hasCycles() {
```

Comment:
Return simple data type `boolean` here.
Furthermore documentation for public method is missing